### ISSUE TITLE
solveset mishandles domain

### ISSUE THREAD

**Issue by smichr** [sympy/sympy#12527]
Giving an explicit domain gives no solution:
```
>>> solveset(1/tan(x),x,Interval(1,10))
EmptySet()
```

whereas giving no domain and intersecting the result gives solutions

```
>>> solveset(1/tan(x),x).intersect(Interval(1,10))
{pi/2, 3*pi/2, 5*pi/2}
```

**Comment by kshitij10496** [comment]
In v1.0:
```python
In []: solveset(1/tan(x), x, Interval(1, 10))
Out[]: ImageSet(Lambda(_n, 2*_n*pi + pi/2), Integers()) U ImageSet(Lambda(_n, 2*_n*pi - pi/2), Integers())
```
In `master`:
```python
In []: solveset(1/tan(x), x, S.Reals)
Out[]: Union(ImageSet(Lambda(_n, 2*_n*pi + pi/2), S.Integers), ImageSet(Lambda(_n, 2*_n*pi + 3*pi/2), S.Integers))
```
This result is pretty weird and a type of regression, I guess (From a somewhat incorrect answer to an absolute wrong result). 
We are supposed to take the intersection of the result obtained in the real domain with the `Interval(1, 10)`, aren't we?
@smichr I will try to dig into this.

**Comment by smichr** [comment]
Yes, if a domain is provided, the answer should come back within that domain.

Regarding the answer being wrong...I don't think there is a problem (unless I am missing something): they both give +/- odd multiples of pi/2.

Did you get the same result as me for `solveset(1/tan(x),x,Interval(1,10))
` in master? If so, have a look at the `domain_check` (or whatever it is). I recall seeing some issues there.

**Comment by kshitij10496** [comment]
> Regarding the answer being wrong...I don't think there is a problem (unless I am missing something): they both give +/- odd multiples of pi/2.

I was referring to the difference between the results obtained in v1.0 and master for `solveset(1/tan(x), x, Interval(0, 10))`.
In v1.0, the `ImageSet` is not reduced to a `FiniteSet` while in the current master, an `EmptySet` is being returned.

> Did you get the same result as me for solveset(1/tan(x),x,Interval(1,10)) in master? If so, have a look at the domain_check (or whatever it is). I recall seeing some issues there.

Yes, I was able to reproduce the bug in the master.

As you mentioned, the issue lies in the behaviour of `domain_check` primarily.
`solveset` is able to process the equation to a state where we obtain the desired solution.
However, a `_check` is executed which calls `domain_check` to check whether all the points in the obtained result are valid i.e they belong to the domain of the input function or not.

In the case of `1/tan(x)`, the `domain_check` functionality splits `f` into its arguments, namely `tan(x)` and `-1`. Since, `pi/2` is a singularity of `tan(x)`, it is removed from the solution set.
In the same manner, all other solutions are rejected and an `EmptySet` is returned.

**Comment by oscarbenjamin** [comment]
This is still an issue on current master:
```julia
In [9]: solveset(1/tan(x),x,Interval(1,10))                                                                                                                   
Out[9]: ∅

In [10]: solveset(1/tan(x),x).intersect(Interval(1,10))                                                                                                       
Out[10]: 
⎧π  3⋅π  5⋅π⎫
⎨─, ───, ───⎬
⎩2   2    2 ⎭
```

**Comment by gschintgen** [comment]
The point-by-point `domain_check` can't be done for infinite ImageSets, that's why there are spurious solutions if solveset is called without a finite domain. Relevant code:
https://github.com/sympy/sympy/blob/81e87776507f2f265833ca0da877df7e74190cc3/sympy/solvers/solveset.py#L982-L997

I think I've filed a duplicate: https://github.com/sympy/sympy/issues/17324

For future reference: there's a noteworthy discussion about this in #12530. (Personally I'm also of the opinion that removable singularities are not actual solutions.)

**Comment by czgdp1807** [comment]
The issue is still open for fixes. Please go through the discussion here and in https://github.com/sympy/sympy/pull/12530

**Comment by oscargus** [comment]
But it should be possible to simply insert the expression of the infinite ImageSet to check if the result evaluates to an infinite expression?

```
In [16]: print(solveset(1/tan(x), x))
Union(ImageSet(Lambda(_n, 2*_n*pi + pi/2), Integers), ImageSet(Lambda(_n, 2*_n*pi + 3*pi/2), Integers))

In [17]: print(1/tan(2*n*pi+pi/2))
zoo

In [18]: print(1/tan(2*n*pi+3*pi/2))
zoo
```

**Comment by oscarbenjamin** [comment]
Is the bug not basically just this:
```python
In [17]: from sympy.solvers.solveset import invert_real, invert_complex

In [18]: invert_real(1/x, 0, x)
Out[18]: (x, ∅)

In [19]: invert_complex(1/x, 0, x)
Out[19]: 
⎛1     ⎞
⎜─, {0}⎟
⎝x     ⎠
```
Here `invert_complex` should give the empty set.

**Comment by smichr** [comment]
> Here `invert_complex` should give the empty set.

An inline notes that the complex inversion of powers is not as nuanced as it is in the real case. So the return value is basically just acknowledging that `invert_complex` can't solve it.
